src/nupic/engine/Link.cpp

```diff
@@ -317,19 +354,40 @@ Link::compute()
 {
   NTA_CHECK(initialized_);
 
+  // If first compute during current network run iteration, append src to
+  // circular buffer
+  if (!srcBuffer_.full())
+  {
+    const Array & srcArray = src_->getData();
+    size_t elementCount = srcArray.getCount();
+    auto elementType = srcArray.getType();
+
+    Array array(elementType);
+    srcBuffer_.push_back(array);
+
+    auto & lastElement = srcBuffer_.back();
+    lastElement.allocateBuffer(elementCount);
+    ::memcpy(lastElement.getBuffer(), srcArray.getBuffer(),
+             elementCount * BasicType::getSize(elementType));
+  }
+
   // Copy data from source to destination.
-  // TBD: with zero-copy optimization, we won't do anything,
-  // but that isn't implemented yet.
-  const Array & src = src_->getData();
+  const Array & src = srcBuffer_[0];
   const Array & dest = dest_->getData();
 
-  // TBD: use src offset and src size (only for certain types of links)
   size_t typeSize = BasicType::getSize(src.getType());
   size_t srcSize = src.getCount() * typeSize;
   size_t destByteOffset = destOffset_ * typeSize;
   ::memcpy((char*)(dest.getBuffer()) + destByteOffset, src.getBuffer(), srcSize);
 }
 
+void Link::purgeHead()
```

Comment:
Can you add "buffer" or something to the name so it is more obvious what it is for?